### PR TITLE
Move Build & Test out of docker

### LINKED ISSUE #27: CI Improvements
Hi, I am looking over the CI system and have some questions and proposals.

- What is the update to #6 ? Is there a reason that its blocked?
- Why do we spin up a docker container rather than using the built-in worker environment?
- Why is the CI is specifically configured to only trigger on the main branch?
- Are we interested in linting the code as checks? Is there a standard lint rule that we follow?
- Are we interested in auto-formatting for PRs?
- Is this still true? 

https://github.com/beman-project/exemplar/blob/d9145b52c9072a569d747547edfc7e03a1494809/.ci/docker/ubuntu.Dockerfile#L3-L4

I know a lot of these configurations are carryover-ed from https://github.com/bretbrownjr/zerocode so that some haven't been evaluated against their use in beman-project. I would happy to implement any improvements.

Action Items from discussions below + weekly sync:
- [x] Close #6
- [x] Remove main branch restriction for CI-trigger #28
- [x] Create Periodic Builds #28 
- [x] Implement #34 
- [x] Auto-format on PRs #46 
- [ ] Update to using GitHub native image instead of docker #48

### PR DESCRIPTION
**Pull request by wusatosi**
This PR moves the build and test system in CI out of docker and refactors the CI script.

This PR is functionally the same as head (with minor improvement). 

Reduce CI time from ~2 minutes down to ~45 seconds.

Based on Discussion here: https://discourse.bemanproject.org/t/should-we-use-or-remove-docker-from-beman/228

Closes #27

### CODE REVIEW

File: .github/workflows/ci_tests.yml [hunk 1]
```diff
@@ -12,63 +12,90 @@ on:
...
+        uses: lukka/get-cmake@latest
+        with:
+          cmakeVersion: "~3.25.0"
+          ninjaVersion: "^1.11.1"
+      - name: Print installed softwares
         run: |
-          echo "image-version=$(echo $ImageVersion)" >> "$GITHUB_OUTPUT"
-        working-directory: .
-      - name: modify number of bits to use for ASLR entropy
-        if: ${{ steps.runner-image-version.outputs.ImageVersion }} == '20240310.1.0'
```
Comment (wusatosi) [inline .github/workflows/ci_tests.yml:47]: we no longer have to care about this. Current image version has well past this post.

(Also this condition was erroneous).

File: .github/workflows/ci_tests.yml [hunk 2]
```diff
@@ -12,63 +12,90 @@ on:
...
-      # The next image is supposed to fix this, so if the Ubuntu image has been
-      # updated, this work around is no longer required.
-      - name: get runner image version
-        id: runner-image-version
+      - name: Install Ninja
+        uses: lukka/get-cmake@latest
+        with:
+          cmakeVersion: "~3.25.0"
+          ninjaVersion: "^1.11.1"
+      - name: Print installed softwares
```
Comment (wusatosi) [inline .github/workflows/ci_tests.yml:67]: GitHub Image comes with Clang and GCC, we can setup specific version of them if needed in the future.

**Review by steve-downey — APPROVED** [review]
👍

File: .github/workflows/ci_tests.yml [hunk 3]
```diff
@@ -12,63 +12,90 @@ on:
...
+          cmake --build build --config Release --verbose
+          cmake --build build --config Release --target all_verify_interface_header_sets
+          cmake --install build --config Release --prefix /opt/beman.exemplar
+          find /opt/beman.exemplar -type f
+        env:
+          CC: ${{ matrix.compiler.c }}
+          CXX: ${{ matrix.compiler.cpp }}
+          CMAKE_GENERATOR: "Ninja Multi-Config"
+      - name: Test Release
+        run: ctest --test-dir build -C Release
```
Comment (camio) [inline .github/workflows/ci_tests.yml:85]: Prefer long options for readability.

```suggestion
        run: ctest --test-dir build --build-config Release
```

File: .github/workflows/ci_tests.yml [hunk 4]
```diff
@@ -12,63 +12,90 @@ on:
...
+          cmake --build build --config Debug --verbose
+          cmake --build build --config Debug --target all_verify_interface_header_sets
+          cmake --install build --config Debug --prefix /opt/beman.exemplar
+          find /opt/beman.exemplar -type f
+        env:
+          CC: ${{ matrix.compiler.c }}
+          CXX: ${{ matrix.compiler.cpp }}
+          CMAKE_GENERATOR: "Ninja Multi-Config"
+      - name: Test Release
+        run: ctest --test-dir build -C Debug
```
Comment (camio) [inline .github/workflows/ci_tests.yml:98]:
```suggestion
        run: ctest --test-dir build --build-config Debug
```

File: .github/workflows/ci_tests.yml [hunk 5]
```diff
@@ -12,63 +12,90 @@ on:
...
+            c: gcc
+          - cpp: clang++
+            c: clang
+        cmake_args:
+          - description: "Static default"
+            args: ""
+          - description: "Dynamic default"
+            args: "-DBUILD_SHARED_LIBS=on"
+          - description: "static C++20"
+            args: "-DCMAKE_CXX_STANDARD=20 -DCMAKE_CXX_STANDARD_REQUIRED=on"
```
Comment (camio) [inline .github/workflows/ci_tests.yml:30]: I don't know what `-DCMAKE_CXX_STANDARD_REQUIRED=on` is doing for us here, but I understand this is coming from what was there before.

**Review by camio — APPROVED** [review]
I added some minor comments, but otherwise approve.

File: .github/workflows/ci_tests.yml (hunk 5 above)
Comment (wusatosi) [reply .github/workflows/ci_tests.yml:30]: I do not understand why this is here as well too. Should I remove it?

File: .github/workflows/ci_tests.yml (hunk 3 above)
Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: Thanks!

File: .github/workflows/ci_tests.yml (hunk 4 above)
Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: Thanks!

File: .github/workflows/ci_tests.yml (hunk 5 above)
Comment (camio) [reply .github/workflows/ci_tests.yml:30]: Yeah, go for it.

Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: Done for all occurrences.

File: .github/workflows/ci_tests.yml [hunk 6]
```diff
@@ -12,63 +12,90 @@ on:
...
-          - { id: ubuntu-clang-static, platform: ubuntu, cc: clang, cpp: clang++, cmake_args: ""}
-          - { id: ubuntu-clang-dynamic, platform: ubuntu, cc: clang, cpp: clang++, cmake_args: "-DBUILD_SHARED_LIBS=on"}
-          - { id: ubuntu-gcc-static-cxx20, platform: ubuntu, cc: gcc, cpp: g++, cmake_args: "-DCMAKE_CXX_STANDARD=20 -DCMAKE_CXX_STANDARD_REQUIRED=on"}
-          - { id: ubuntu-gcc-static-cxx23, platform: ubuntu, cc: gcc, cpp: g++, cmake_args: "-DCMAKE_CXX_STANDARD=23 -DCMAKE_CXX_STANDARD_REQUIRED=on"}
-          - { id: ubuntu-gcc-static-cxx26, platform: ubuntu, cc: gcc, cpp: g++, cmake_args: "-DCMAKE_CXX_STANDARD=26 -DCMAKE_CXX_STANDARD_REQUIRED=on"}
-          - { id: ubuntu-clang-static-cxx20, platform: ubuntu, cc: clang, cpp: clang++, cmake_args: "-DCMAKE_CXX_STANDARD=20 -DCMAKE_CXX_STANDARD_REQUIRED=on"}
-          - { id: ubuntu-clang-static-cxx23, platform: ubuntu, cc: clang, cpp: clang++, cmake_args: "-DCMAKE_CXX_STANDARD=23 -DCMAKE_CXX_STANDARD_REQUIRED=on"}
-          - { id: ubuntu-clang-static-cxx26, platform: ubuntu, cc: clang, cpp: clang++, cmake_args: "-DCMAKE_CXX_STANDARD=26 -DCMAKE_CXX_STANDARD_REQUIRED=on"}
```
Comment (neatudarius) [inline .github/workflows/ci_tests.yml:31]: Why do we remove this matrix? We still need to run on with multiple possible compilers x standard versions.

Comment (wusatosi) [reply .github/workflows/ci_tests.yml:31]: This is not removed, it's refactored. 
See code below.

The generated matrix is the same.

Comment (neatudarius) [reply .github/workflows/ci_tests.yml:31]: My bad. Sorry.

**Review by neatudarius — APPROVED** [review]
LGTM 👍

File: .github/workflows/ci_tests.yml (hunk 5 above)
Comment (wusatosi) [reply .github/workflows/ci_tests.yml]: David, can you double check my updated readme file? Otherwise this PR should be ready to be merged.

**Pull request merged**